src/engine/BMTournament.php

```diff
@@ -0,0 +1,688 @@
+     */
+    protected function update_tournament_state_cancelled() {
+    }
+
+    /**
+     * Carry out all automated tournament actions until a player needs to do something
+     *
+     * The idea here is to call update_tournament_state() and do_next_step() one after
+     * another until:
+     * (i) the tournament requires a player action, or
+     * (ii) the tournament reaches a tournament state that signals that it has ended
+     *
+     * The variable $tournamentStateBreakpoint is used for debugging purposes only.
+     * If used, the tournament will stop as soon as the tournament state becomes
+     * the value of $tournamentStateBreakpoint.
+     *
+     * @param int $tournamentStateBreakpoint
+     */
+    public function proceed_to_next_user_action($tournamentStateBreakpoint = NULL) {
+        $repeatCount = 0;
+        $initialTournamentState = $this->tournamentState;
+        $this->update_tournament_state();
+
+        if (isset($tournamentStateBreakpoint) &&
+            ($tournamentStateBreakpoint == $this->tournamentState) &&
+            ($initialTournamentState != $this->tournamentState)) {
+            return;
+        }
+
+        $this->do_next_step();
+
+        while (!$this->isWaitingOnAnyAction()) {
+            $tempTournamentState = $this->tournamentState;
+            $this->update_tournament_state();
+
+            if (isset($tournamentStateBreakpoint) &&
+                ($tournamentStateBreakpoint == $this->tournamentState)) {
+                return;
+            }
+
+            $this->do_next_step();
+
+            if ($this->tournamentState >= BMTournamentState::END_TOURNAMENT) {
+                return;
+            }
+
+            if ($tempTournamentState === $this->tournamentState) {
+                $repeatCount++;
+            } else {
+                $repeatCount = 0;
+            }
+            if ($repeatCount >= 20) {
+                $finalTournamentState = BMTournamentState::as_string($this->tournamentState);
+                throw new LogicException(
+                    'Infinite loop detected when advancing tournament state. '.
+                    "Final tournament state: $finalTournamentState"
+                );
+            }
+        }
+    }
+
+    /**
+     * Is the tournament waiting on any player actions?
+     *
+     * @return bool
+     */
+    protected function isWaitingOnAnyAction() {
+        if (count($this->playerIdArray) < $this->nPlayers) {
+            return TRUE;
+        }
+
+        if (BMTournamentState::PLAY_GAMES == $this->tournamentState) {
+            // check whether all games in this current tournament round are complete
+            foreach ($this->gameIdArrayArray[$this->roundNumber - 1] as $gameId) {
+                $interface = new BMInterface($this->isTest);
+                if (!$interface->game()->isGameComplete($gameId)) {
+                    return TRUE;
+                }
+            }
+        }
+
+        return FALSE;
+    }
```

Comment:
I think I've addressed this with the fix to the above comment where I'm now throwing a LogicException when a player hasn't selected a button.

I was thinking that it might potentially be possible to have no specified button with random button selection, but I've tested this and random button selection is resolved on joining the tournament. Obviously, past me was smart enough to account for me.